website_mrp_bom/tests/test_website_mrp_bom.py

```diff
@@ -0,0 +1,114 @@
+            'product_tmpl_id', '=', cls.product_no_sale.id
+        )], limit=1)
+
+        cls.bom_model = cls.env['mrp.bom']
+        cls.bom_published_1 = cls.bom_model.create({
+            'product_tmpl_id': cls.product_sale_ok.id,
+            'product_id': cls.product_product_sale_ok.id,
+            'product_qty': 1,
+            'type': 'normal',
+            'website_published': True,
+        })
+
+        cls.bom_line_model_1 = cls.env['mrp.bom.line']
+        cls.bom_line_model_1.create({
+            'bom_id': cls.bom_published_1.id,
+            'product_id': cls.product_product_no_sale.id,
+            'product_qty': 2,
+        })
+
+        cls.bom_published_2 = cls.bom_model.create({
+            'product_tmpl_id': cls.product_no_sale.id,
+            'product_id': cls.product_product_no_sale.id,
+            'product_qty': 1,
+            'type': 'normal',
+        })
+        cls.bom_published_2.write({'website_published': True})
+
+        cls.bom_line_model_2 = cls.env['mrp.bom.line']
+        cls.bom_line_model_2.create({
+            'bom_id': cls.bom_published_2.id,
+            'product_id': cls.product_product_sale_ok.id,
+            'product_qty': 2,
+        })
+
+    def test_compute_website_url(self):
+        """Test website_url from mrp.bom."""
+        self.assertTrue('/mrp_bom/' in self.bom_published_1.website_url)
+        self.assertTrue('/mrp_bom/' in self.bom_published_2.website_url)
+
+    @patch('odoo.addons.website_mrp_bom.'
+           'controllers.main.request')
+    @patch('odoo.http.request')
+    def test_request_mrp_bom_1(self, request, request2):
+        """Test controller for mrp_bom for a product salable with
+        mrp_bom_line product not salable"""
+        # Mock
+        request.env = self.env
+        request2.env = self.env
+
+        ctrl = WebsiteMrpBomPage()
+        response = ctrl.boms_detail(self.bom_published_1)
+        self.assertTrue('request.render()' in str(response.name))
+
+    @patch('odoo.addons.website_mrp_bom.'
+           'controllers.main.request')
+    @patch('odoo.http.request')
+    def test_request_mrp_bom_2(self, request, request2):
+        """Test controller for mrp_bom for a product not salable with
+                mrp_bom_line product salable"""
+        # Mock
+        request.env = self.env
+        request2.env = self.env
+
+        ctrl = WebsiteMrpBomPage()
+        response = ctrl.boms_detail(self.bom_published_2)
+        self.assertTrue('request.render()' in str(response.name))
+
+    @patch('odoo.addons.website_mrp_bom.'
+           'controllers.main.request')
+    @patch('odoo.http.request')
+    def test_request_mrp_no_bom(self, request, request2):
+        """Test controller for non existing mrp_bom """
+        # Mock
+        request.env = self.env
+        request2.env = self.env
+
+        ctrl = WebsiteMrpBomPage()
+        response = ctrl.boms_detail(0)
+        self.assertTrue('request.not_found()' in str(response.name))
```

Comment:
you should check that proper exception is raised

Reply:
Using ``` isinstance(bom, request.env['mrp.bom'].__class__) ``` instead of ```bom.exists()``` don't make any exception run, if model is not correctly put on URL, it will render not found page.